internal/kafka/internal/services/data_plane_kafka.go

```diff
@@ -52,67 +67,128 @@ func NewDataPlaneKafkaService(kafkaSrv KafkaService, clusterSrv ClusterService,
-			// Store the routes (and create them) if they are available at this stage to lessen the length of time taken to provision the Kafka.
-			// The routes list will either be empty or complete.
-			e = d.persistKafkaRoutes(kafka, ks, cluster)
-		case statusError:
-			// when getStatus returns statusError we know that the ready
-			// condition will be there so there's no need to check for it
-			readyCondition, _ := ks.GetReadyCondition()
-			e = d.setKafkaClusterFailed(kafka, readyCondition.Message)
-		case statusDeleted:
-			e = d.setKafkaClusterDeleting(kafka)
-		case statusRejected:
-			e = d.reassignKafkaCluster(kafka)
-		case statusRejectedClusterFull:
-			e = d.unassignKafkaFromDataplaneCluster(kafka)
-		case statusUnknown:
-			log.Infof("kafka cluster %s status is unknown", ks.KafkaClusterId)
-		default:
-			log.V(5).Infof("kafka cluster %s is still installing", ks.KafkaClusterId)
+
+	prewarmingStatusInfo := reservedManagedKafkaStatusCountPerInstanceType{}
+	supportedInstanceTypes := cluster.GetSupportedInstanceTypes()
+
+	// prefill prewarming status info with empty count
+	for _, supportedInstanceType := range supportedInstanceTypes {
+		// sets all the statuses to 0 to allow for the count to drop
+		prewarmingStatusInfo[api.ClusterInstanceTypeSupport(supportedInstanceType)] = totalCountPerReservedManagedKafkaDeploymentStatus{
+			statusReady:               0,
+			statusError:               0,
+			statusUnknown:             0,
+			statusDeleted:             0,
+			statusRejected:            0,
+			statusInstalling:          0,
+			statusRejectedClusterFull: 0,
 		}
-		if e != nil {
-			log.Error(errors.Wrapf(e, "Error updating kafka %s status", ks.KafkaClusterId))
+	}
+
+	for _, ks := range status {
+		managedKafkaDeploymentType := getManagedKafkaDeploymentType(ks)
+		switch managedKafkaDeploymentType {
+		case realDeploymentType:
+			d.processRealKafkaDeployment(ks, cluster, log)
+		case reservedDeploymentType:
+			d.processReservedKafkaDeployment(ks, prewarmingStatusInfo, log, clusterID)
 		}
+	}
 
-		e = d.setKafkaRequestVersionFields(kafka, ks)
-		if e != nil {
-			log.Error(errors.Wrapf(e, "Error updating kafka '%s' version fields", ks.KafkaClusterId))
+	// emits the kas_fleet_manager_prewarmed_kafka_instances metrics
+	for instanceType, instanceTypePrewarmingStatusInfo := range prewarmingStatusInfo {
+		for status, count := range instanceTypePrewarmingStatusInfo {
+			metrics.UpdateClusterPrewarmingStatusInfoCountMetric(metrics.PrewarmingStatusInfo{
+				ClusterID:    cluster.ClusterID,
+				InstanceType: instanceType.String(),
+				Status:       status.String(),
+				Count:        count,
+			})
 		}
 	}
+
 	return nil
 }
 
+// processRealKafkaDeployment process reserved kafka instances and updates their status in cluster
+func (d *dataPlaneKafkaService) processReservedKafkaDeployment(ks *dbapi.DataPlaneKafkaStatus, prewarmingStatusInfo reservedManagedKafkaStatusCountPerInstanceType, log logger.UHCLogger, clusterID string) {
```

Comment:
A nice side-effect of the refactor is that no DB errors will be shown anymore when processing the reserved managed kafkas :+1: